### DISCUSSION TITLE
Support virtual workspaces

### DISCUSSION THREAD

**Discussion by lszomoru** [microsoft/pylance-release#2696] [Enhancement]
👋 Hi there, Martin here, from the VS Code team.

Recently we've announced the [Remote Repository feature](https://code.visualstudio.com/updates/v1_56#_remote-repositories-remotehub) that lets you browse and edit files and folders directly on GitHub.

`Open Remote Repository...` opens VSCode on a folder or workspace located on a virtual file system. We call this a **virtual workspace**. We observed that not all extension support this well, either because they can not, or they haven't thought about it.

It would be fantastic if you could test whether your extension can handle virtual workspaces:

Check out the [Virtual Workspaces Extension Author Guide](https://github.com/microsoft/vscode/wiki/Virtual-Workspaces) on how to do that.

When done, set the new `virtualWorkspaces` capability in your 'package.json'.

```
{
  "capabilities": {
    "virtualWorkspaces": true | false
  }
}
```

- Use `"virtualWorkspaces": true` if your extension is prepared for virtual workspaces
- Use `"virtualWorkspaces": false` if your extension should be disabled when a virtual workspace is opened

For questions and comments please use the [Virtual Workspaces Tracking Issue](https://github.com/microsoft/vscode/issues/123115).

Thanks for the support and the great work! ❤️

**Comment by jakebailey** [comment]
Pylance is a language server and runs outside of VS Code, so we can't really use any of these APIs. This is about the same as the VFS that's used for SSH in some conditions and likely will cause the same breakages (because the LS has no idea that VS Code has its own filesystem abstraction LSs have no accesses to).

I'm not sure how we can support this without entirely running our FS accesses through a custom LSP-based API. These sorts of things are a bit unfortunate when everyone is moving to language servers...

**Comment by aeschli** [comment]
It's completely fine to only provide limited functionality or disable the extension.
- The main goal is to that we don't get any error dialogs when virtual resources are opened.
- The language extensions that ship with VSCode (TypeScript, JSON, CSS, HTML, Markdown) just look at the open text document ('single file only').

**Comment by DevilXD** [comment]
The Python extension is currently disabled on virtual workspaces too, which makes it impossible to use Pylance on those anyway: https://i.imgur.com/mpBCf3M.png

**Comment by jakebailey** [comment]
This should now be partially supported for open files only, as of the last couple of releases. Getting more than this will be challenging.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
